protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,509 @@
+///    # use libp2p_relay::v2::client;
+///    # let inner_transport = MemoryTransport::default();
+///    # let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+///    #     PeerId::random(),
+///    #     inner_transport,
+///    # );
+///    let dst_addr_via_relay = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(PeerId::random().into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit) // Signal to connect via relay and not directly.
+///        .with(Protocol::Memory(42)) // Destination address.
+///        .with(Protocol::P2p(PeerId::random().into())); // Destination peer id.
+///    transport.dial(dst_addr_via_relay).unwrap();
+///    ```
+///
+/// 3. Listen for incoming relayed connections via specific relay.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_relay::v2::client;
+///    # let inner_transport = MemoryTransport::default();
+///    # let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+///    #     PeerId::random(),
+///    #     inner_transport,
+///    # );
+///    let relay_addr = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(PeerId::random().into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit); // Signal to listen via remote relay node.
+///    transport.listen_on(relay_addr).unwrap();
+///    ```
+#[derive(Clone)]
+pub struct ClientTransport<T: Clone> {
+    to_behaviour: mpsc::Sender<TransportToBehaviourMsg>,
+
+    inner_transport: T,
+}
+
+impl<T: Clone> ClientTransport<T> {
+    /// Create a new [`ClientTransport`] by wrapping an existing [`Transport`] in a
+    /// [`ClientTransport`].
+    ///
+    ///```
+    /// # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+    /// # use libp2p_core::transport::memory::MemoryTransport;
+    /// # use libp2p_relay::v2::client;
+    /// let inner_transport = MemoryTransport::default();
+    /// let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+    ///    PeerId::random(),
+    ///    inner_transport,
+    /// );
+    ///```
+    pub(crate) fn new(t: T) -> (Self, mpsc::Receiver<TransportToBehaviourMsg>) {
+        let (to_behaviour, from_transport) = mpsc::channel(0);
+
+        let transport = ClientTransport {
+            to_behaviour,
+
+            inner_transport: t,
+        };
+
+        (transport, from_transport)
+    }
+}
+
+impl<T: Transport + Clone> Transport for ClientTransport<T> {
+    type Output = EitherOutput<<T as Transport>::Output, RelayedConnection>;
+    type Error = EitherError<<T as Transport>::Error, RelayError>;
+    type Listener = RelayListener<T>;
+    type ListenerUpgrade = RelayedListenerUpgrade<T>;
+    type Dial = EitherFuture<<T as Transport>::Dial, RelayedDial>;
```

Comment:
```suggestion
    type Output = EitherOutput<T::Output, RelayedConnection>;
    type Error = EitherError<T::Error, RelayError>;
    type Listener = RelayListener<T>;
    type ListenerUpgrade = RelayedListenerUpgrade<T>;
    type Dial = EitherFuture<T::Dial, RelayedDial>;
```
No need for fully-qualified syntax here.

Reply:
Good point.

Fixed via using `OrTransport` instead. See a8f21ae7.